examples/psp22-extension/runtime/psp22-extension-example.rs

```diff
@@ -0,0 +1,398 @@
+use codec::{
+    Decode,
+    Encode,
+    MaxEncodedLen,
+};
+use frame_support::{
+    dispatch::RawOrigin,
+    log::{
+        error,
+        trace,
+    },
+    pallet_prelude::*,
+    traits::fungibles::{
+        approvals::{
+            Inspect as AllowanceInspect,
+            Mutate as AllowanceMutate,
+        },
+        Inspect,
+        InspectMetadata,
+        Transfer,
+    },
+};
+use pallet_assets::{
+    self,
+    WeightInfo,
+};
+use pallet_contracts::chain_extension::{
+    ChainExtension,
+    Environment,
+    Ext,
+    InitState,
+    RetVal,
+    SysConfig,
+    UncheckedFrom,
+};
+use sp_runtime::{
+    traits::{
+        Saturating,
+        StaticLookup,
+        Zero,
+    },
+    DispatchError,
+};
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22BalanceOfInput<AssetId, AccountId> {
+    pub asset_id: AssetId,
+    pub owner: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22AllowanceInput<AssetId, AccountId> {
+    pub asset_id: AssetId,
+    pub owner: AccountId,
+    pub spender: AccountId,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub to: AccountId,
+    pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22TransferFromInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub from: AccountId,
+    pub to: AccountId,
+    pub value: Balance,
+}
+
+#[derive(Debug, PartialEq, Encode, Decode, MaxEncodedLen)]
+struct Psp22ApproveInput<AssetId, AccountId, Balance> {
+    pub asset_id: AssetId,
+    pub spender: AccountId,
+    pub value: Balance,
+}
+
+pub struct Psp22Extension;
+
+fn map_err(err_msg: &'static str) -> impl FnOnce(DispatchError) -> DispatchError {
+    move |err| {
+        trace!(
+            target: "runtime",
+            "PSP22 Transfer failed:{:?}",
+            err
+        );
+        DispatchError::Other(err_msg)
+    }
+}
+
+fn metadata<T, E>(
+    func_id: u32,
+    env: Environment<E, InitState>,
+) -> Result<(), DispatchError>
+where
+    T: pallet_assets::Config + pallet_contracts::Config,
+    <T as SysConfig>::AccountId: UncheckedFrom<<T as SysConfig>::Hash> + AsRef<[u8]>,
+    E: Ext<T = T>,
+{
+    let mut env = env.buf_in_buf_out();
+    let asset_id = env.read_as()?;
+    let result = match func_id {
```

Comment:
We can `encode()` the `result` once instead of at every match arm

Reply:
Can you address this for the other functions too?

Reply:
I can do this in the `query` function, but here it won't work because of incompatible match arms (`name` and `symbol` return `Vec<u8>` while `decimals` returns `u8`).